### PR TITLE
Added other sampling techniques

### LINKED ISSUE #73: Try other techniques to handle imbalanced datasets
For now we're simply randomly under-sampling the majority classes.

### PR DESCRIPTION
**Pull request by Aftaab99**
Fixes #73 
Added BorderlineSMOTE and Random bootstrap sampling methods. 

Cross validation scores
1) Undersampling:
`Accuracy: f0.936495609359931 (+/- 0.022820414517239145)`
`Precision: f0.9519224345403131 (+/- 0.012983693728434824)`
`Recall: f0.9193644992639968 (+/- 0.035993231013244996)` 

2) BordelineSMOTE oversampling:
`Accuracy: f0.9433541773895453 (+/- 0.013859958420733072)`
`Precision: f0.9811981372337684 (+/- 0.0064406491621750205)`
`Recall: f0.9503288860002288 (+/- 0.01694315771413166)`

3) RandomOverSampling:
`Accuracy: f0.9343318950396726 (+/- 0.011400760695446618)`
`Precision: f0.9862632613310028 (+/- 0.005463803624684358)`
`Recall: f0.9343066513407194 (+/- 0.012411187151863524)`

I could not execute the test scores part as I have less RAM and was getting memory errors. So I'm not sure if this will reflect on the test scores.

### CODE REVIEW

**Comment by marco-c** [comment]
Which `goal` did you select when running run.py?
Can you try the same with the `tracking` goal?

Did you try all methods from imblearn, or are there other methods to try?

**Comment by saurabhdaalia** [comment]
@Aftaab99  @marco-c  , we can also try `ADASYN (Adaptive synthetic sampling approach)` or` SMOTEENN`, `SMOTETomek` techniques available in imblearn.

**Comment by Aftaab99** [comment]
> Which `goal` did you select when running run.py?

It was for the `bug` model.
> Can you try the same with the `tracking` goal?

I did try them but again I could only compute the cross-validation accuracy after which I run into a `std::bad_alloc()` memory error. The ones I did previously were oversampling methods and they increase the size of the train data quite a bit, making it hard to fit into memory.

> Did you try all methods from imblearn, or are there other methods to try?

I did. I could compute test scores for only 3 methods as others ran out of memory.
All of these are for the `bug` model.

1. Current method - Random Undersampling
```
X: (6651, 242246), y: (6651,)
X_train: (1984, 242246), y_train: (1984,)
X_test: (666, 242246), y_test: (666,)
Cross Validation scores:
Accuracy: f0.936495609359931 (+/- 0.022820414517239145)
Precision: f0.9519224345403131 (+/- 0.012983693728434824)
Recall: f0.9193644992639968 (+/- 0.035993231013244996)

Top 26 Features :
1. 'comments__regression' (-0.07241272926330566)
2. 'data__severity=critical' (+0.044457025825977325)
3. 'data__ever_affected=False' (+0.043717965483665466)
4. 'first_comment__mozilla' (+0.04288704693317413)
5. 'first_comment__str' (+0.042344532907009125)
6. 'first_comment__expected' (+0.03356219828128815)
7. 'comments__nightly' (-0.03281090781092644)
8. 'first_comment__10' (+0.03086252324283123)
9. 'comments__issue' (+0.030534988269209862)
10. 'comments__reproduce' (+0.029530344530940056)
11. 'comments__marked' (+0.026480328291654587)
12. 'first_comment__regression' (+0.026208793744444847)
13. 'comments__remove' (-0.024171311408281326)
14. 'comments__pushlog' (+0.0237937793135643)
15. 'first_comment__attachment' (+0.023006154224276543)
16. 'comments__fix' (-0.022536173462867737)
17. 'comments__use' (-0.018688738346099854)
18. 'first_comment____code_reference_url__' (+0.018605975434184074)
19. 'comments__57' (+0.01682201214134693)
20. 'title__crash' (+0.0157783105969429)
21. 'comments__work' (-0.015536600723862648)
22. 'comments__support' (+0.01366322673857212)
23. 'first_comment__build' (+0.012840716168284416)
24. 'comments__click' (+0.012430374510586262)
25. 'first_comment__actual' (+0.011866588145494461)
26. 'comments__add' (-0.011244802735745907)
No confidence threshold - 666 classified
[[500  49]
 [  8 109]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.91      0.93      0.95      0.92      0.85       549
          0       0.69      0.93      0.91      0.79      0.92      0.85       117

avg / total       0.93      0.91      0.93      0.92      0.92      0.85       666


Confidence threshold > 0.6 - 632 classified
[[487  37]
 [  4 104]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.93      0.96      0.96      0.95      0.89       524
          0       0.74      0.96      0.93      0.84      0.95      0.90       108

avg / total       0.95      0.94      0.96      0.94      0.95      0.89       632


Confidence threshold > 0.7 - 601 classified
[[471  27]
 [  2 101]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.95      0.98      0.97      0.96      0.92       498
          0       0.79      0.98      0.95      0.87      0.96      0.93       103

avg / total       0.96      0.95      0.97      0.95      0.96      0.93       601


Confidence threshold > 0.8 - 569 classified
[[457  16]
 [  2  94]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.97      0.98      0.98      0.97      0.94       473
          0       0.85      0.98      0.97      0.91      0.97      0.95        96

avg / total       0.97      0.97      0.98      0.97      0.97      0.95       569


Confidence threshold > 0.9 - 500 classified
[[414   8]
 [  1  77]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.98      0.99      0.99      0.98      0.97       422
          0       0.91      0.99      0.98      0.94      0.98      0.97        78

avg / total       0.98      0.98      0.99      0.98      0.98      0.97       500
``` 

2) Condensed Nearest neighbours(worse CV accuracy much best test metrics)
```
X: (6651, 242246), y: (6651,)
X_train: (2130, 242246), y_train: (2130,)
X_test: (666, 242246), y_test: (666,)
Cross Validation scores:
Accuracy: f0.896212812716867 (+/- 0.030623577490069716)
Precision: f0.9163618178917126 (+/- 0.021561124529853492)
Recall: f0.8866063838009121 (+/- 0.0432212111543417)

Top 26 Features :
1. 'comments__str' (-0.07270163297653198)
2. 'comments__regression' (+0.06828715652227402)
3. 'first_comment__expected' (+0.06344452500343323)
4. 'first_comment__str' (-0.04234915226697922)
5. 'comments__reproduce' (+0.040977898985147476)
6. 'comments__fix' (+0.026220684871077538)
7. 'comments__click' (+0.024248404428362846)
8. 'title__remove' (-0.024124210700392723)
9. 'comments__remove' (-0.023561982437968254)
10. 'first_comment__regression' (+0.022430771961808205)
11. 'comments__nightly' (+0.02241024374961853)
12. 'comments__57' (+0.02130780555307865)
13. 'first_comment__error' (+0.020929714664816856)
14. 'comments__duplicate' (+0.020234905183315277)
15. 'first_comment__attachment' (+0.019311539828777313)
16. 'comments__marked' (+0.017358627170324326)
17. 'comments__want' (+0.015838703140616417)
18. 'comments__pushlog' (+0.015176941640675068)
19. 'first_comment__mozilla' (+0.014647424221038818)
20. 'title____file_reference__' (-0.014067177660763264)
21. 'comments__10' (+0.013351477682590485)
22. 'first_comment__nightly' (+0.013020610436797142)
23. 'first_comment____url__' (+0.012839407660067081)
24. 'first_comment__11' (+0.011986522004008293)
25. 'title__add' (-0.01196298748254776)
26. 'comments__make' (-0.011323346756398678)
No confidence threshold - 666 classified
[[519  30]
 [ 13 104]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.95      0.89      0.96      0.92      0.85       549
          0       0.78      0.89      0.95      0.83      0.92      0.84       117

avg / total       0.94      0.94      0.90      0.94      0.92      0.84       666


Confidence threshold > 0.6 - 624 classified
[[497  18]
 [  7 102]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.97      0.94      0.98      0.95      0.91       515
          0       0.85      0.94      0.97      0.89      0.95      0.90       109

avg / total       0.96      0.96      0.94      0.96      0.95      0.90       624


Confidence threshold > 0.7 - 584 classified
[[474  11]
 [  5  94]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.98      0.95      0.98      0.96      0.93       485
          0       0.90      0.95      0.98      0.92      0.96      0.93        99

avg / total       0.97      0.97      0.95      0.97      0.96      0.93       584


Confidence threshold > 0.8 - 516 classified
[[436   4]
 [  3  73]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.99      0.96      0.99      0.98      0.95       440
          0       0.95      0.96      0.99      0.95      0.98      0.95        76

avg / total       0.99      0.99      0.97      0.99      0.98      0.95       516


Confidence threshold > 0.9 - 426 classified
[[380   1]
 [  2  43]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      1.00      0.96      1.00      0.98      0.96       381
          0       0.98      0.96      1.00      0.97      0.98      0.95        45


avg / total       0.99      0.99      0.96      0.99      0.98      0.96       426

```

3) Increased Hardness Threshold(Best CV accuracy, bad test scores)
```
X: (6651, 242246), y: (6651,)
X_train: (2242, 242246), y_train: (2242,)
X_test: (666, 242246), y_test: (666,)
Cross Validation scores:
Accuracy: f0.9866170855870188 (+/- 0.01322850224529516)
Precision: f0.9841785711394913 (+/- 0.015604330138386833)
Recall: f0.992 (+/- 0.015178932768808233)

Top 16 Features :
1. 'first_comment__10' (-0.11646901071071625)
2. 'data__severity=critical' (+0.09525153040885925)
3. 'comments__pushlog' (+0.05952067673206329)
4. 'comments__regression' (+0.047277260571718216)
5. 'first_comment__mozilla' (+0.045746054500341415)
6. 'first_comment__str' (+0.03813517466187477)
7. 'comments__nightly' (-0.037752848118543625)
8. 'comments__use' (-0.031280823051929474)
9. 'comments__10' (+0.030948081985116005)
10. 'first_comment__build' (+0.029887259006500244)
11. 'first_comment__regression' (-0.028702251613140106)
12. 'comments__reproduce' (+0.025912268087267876)
13. 'comments__actual' (-0.021665671840310097)
14. 'first_comment__filed' (+0.021482082083821297)
15. 'data__ever_affected=False' (-0.01942470110952854)
16. 'first_comment__actual' (-0.018398305401206017)
No confidence threshold - 666 classified
[[424 125]
 [  2 115]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.77      0.98      0.87      0.87      0.74       549
          0       0.48      0.98      0.77      0.64      0.87      0.78       117

avg / total       0.90      0.81      0.95      0.83      0.87      0.75       666


Confidence threshold > 0.6 - 646 classified
[[414 115]
 [  2 115]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.78      0.98      0.88      0.88      0.75       529
          0       0.50      0.98      0.78      0.66      0.88      0.78       117

avg / total       0.91      0.82      0.95      0.84      0.88      0.76       646


Confidence threshold > 0.7 - 617 classified
[[404  97]
 [  1 115]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.81      0.99      0.89      0.89      0.78       501
          0       0.54      0.99      0.81      0.70      0.89      0.81       116

avg / total       0.91      0.84      0.96      0.86      0.89      0.79       617


Confidence threshold > 0.8 - 583 classified
[[382  86]
 [  1 114]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.82      0.99      0.90      0.90      0.79       468
          0       0.57      0.99      0.82      0.72      0.90      0.82       115

avg / total       0.91      0.85      0.96      0.86      0.90      0.80       583


Confidence threshold > 0.9 - 539 classified
[[357  70]
 [  1 111]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.84      0.99      0.91      0.91      0.82       427
          0       0.61      0.99      0.84      0.76      0.91      0.84       112

avg / total       0.92      0.87      0.96      0.88      0.91      0.82       539
```

4) Near miss undersampling (was neither good at CV nor test)
```
X: (6651, 242246), y: (6651,)
X_train: (1984, 242246), y_train: (1984,)
X_test: (666, 242246), y_test: (666,)
Cross Validation scores:
Accuracy: f0.9017181868940662 (+/- 0.04442253900462353)
Precision: f0.9310648069202795 (+/- 0.023408933295685736)
Recall: f0.8679457895538297 (+/- 0.09655339599109533)

Top 24 Features :
1. 'comments__regression' (+0.08106156438589096)
2. 'first_comment__str' (-0.07494206726551056)
3. 'first_comment__expected' (+0.05551895871758461)
4. 'comments__str' (-0.04094623401761055)
5. 'data__patches' (-0.032244011759757996)
6. 'comments__need' (-0.028087923303246498)
7. 'comments__reproduce' (+0.02692250721156597)
8. 'comments__fix' (-0.025835901498794556)
9. 'data__blocked_bugs_number' (-0.02549518272280693)
10. 'title__remove' (-0.02282004989683628)
11. 'comments__support' (-0.020667975768446922)
12. 'comments__0a1' (+0.020019710063934326)
13. 'comments__remove' (+0.01983777992427349)
14. 'first_comment__regression' (+0.01983177848160267)
15. 'comments__click' (+0.01982012204825878)
16. 'comments__fixed' (+0.019277604296803474)
17. 'comments__error' (+0.018976064398884773)
18. 'comments__approved' (-0.018229015171527863)
19. 'comments__duplicate' (+0.017786553129553795)
20. 'comments__want' (-0.01747107319533825)
21. 'comments__attachment' (+0.016488201916217804)
22. 'comments__marked' (+0.01574762724339962)
23. 'comments__performance' (-0.014025028795003891)
24. 'title____file_reference__' (-0.01297345943748951)
No confidence threshold - 666 classified
[[437 112]
 [ 11 106]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.80      0.91      0.88      0.85      0.71       549
          0       0.49      0.91      0.80      0.63      0.85      0.73       117

avg / total       0.89      0.82      0.89      0.83      0.85      0.72       666


Confidence threshold > 0.6 - 611 classified
[[410  91]
 [  9 101]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.82      0.92      0.89      0.87      0.74       501
          0       0.53      0.92      0.82      0.67      0.87      0.76       110

avg / total       0.90      0.84      0.90      0.85      0.87      0.75       611


Confidence threshold > 0.7 - 532 classified
[[371  59]
 [  6  96]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.86      0.94      0.92      0.90      0.81       430
          0       0.62      0.94      0.86      0.75      0.90      0.82       102

avg / total       0.91      0.88      0.93      0.89      0.90      0.81       532


Confidence threshold > 0.8 - 435 classified
[[322  29]
 [  4  80]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.92      0.95      0.95      0.93      0.87       351
          0       0.73      0.95      0.92      0.83      0.93      0.88        84

avg / total       0.94      0.92      0.95      0.93      0.93      0.87       435


Confidence threshold > 0.9 - 320 classified
[[254   5]
 [  1  60]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.98      0.98      0.99      0.98      0.96       259
          0       0.92      0.98      0.98      0.95      0.98      0.96        61

avg / total       0.98      0.98      0.98      0.98      0.98      0.96       320
```

So which method do we use? And which accuracy are we trying to improve, the CV accuracy or test precision, recall, f1?

**Comment by marco-c** [comment]
To avoid the memory errors, try to add `min_df=0.001` to the text_vectorizer calls in bugbug/model/tracking.py.

**Comment by marco-c** [comment]
first try with 0.0001, then 0.001

File: bugbug/model.py [hunk 1]
```diff
@@ -24,7 +27,9 @@ def __init__(self, lemmatization=False):
         else:
             self.text_vectorizer = TfidfVectorizer
 
-        self.undersampling_enabled = True
+        self.undersampling_enabled = False
```
Comment (marco-c) [inline bugbug/model.py]: Instead of having variables such as this, let's see if we can make each model decide what undersampling/oversampling to use directly in the model implementation (e.g. by adding the sampler to the pipeline for each model, or by having a `self.sampler` property for each model).

**Comment by Aftaab99** [comment]
`BorderlineSMOTE` does better than the current random undersampling for the `bug` model.
`InstanceHardnessThreshold` gives the `tracking` model somewhat better precision.
I've also added the sampler as the sampler as property in the model itself like you had asked.

**Comment by Aftaab99** [comment]
Scores for **Bug model**
i) Random Undersampling
```
Cross Validation scores:
Accuracy: f0.9228011113166769 (+/- 0.016904848046034645)
Precision: f0.9902539192466694 (+/- 0.003719907438735845)
Recall: f0.9164790040541544 (+/- 0.018195026548789843)
X_train: (1984, 30370), y_train: (1984,)
X_test: (666, 30370), y_test: (666,)

Test:
No confidence threshold - 666 classified
[[498  51]
 [  7 110]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.91      0.94      0.94      0.92      0.85       549
          0       0.68      0.94      0.91      0.79      0.92      0.86       117

avg / total       0.93      0.91      0.93      0.92      0.92      0.85       666


Confidence threshold > 0.6 - 629 classified
[[484  38]
 [  3 104]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.93      0.97      0.96      0.95      0.90       522
          0       0.73      0.97      0.93      0.84      0.95      0.91       107

avg / total       0.95      0.93      0.96      0.94      0.95      0.90       629


Confidence threshold > 0.7 - 600 classified
[[472  27]
 [  2  99]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.95      0.98      0.97      0.96      0.92       499
          0       0.79      0.98      0.95      0.87      0.96      0.93       101

avg / total       0.96      0.95      0.97      0.95      0.96      0.93       600


Confidence threshold > 0.8 - 568 classified
[[453  21]
 [  1  93]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.96      0.99      0.98      0.97      0.94       474
          0       0.82      0.99      0.96      0.89      0.97      0.95        94

avg / total       0.97      0.96      0.98      0.96      0.97      0.94       568


Confidence threshold > 0.9 - 499 classified
[[413   6]
 [  1  79]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      0.99      0.99      0.99      0.99      0.97       419
          0       0.93      0.99      0.99      0.96      0.99      0.97        80

```
ii) BorderlineSMOTE
```
Cross Validation scores:
Accuracy: f0.9475336617749791 (+/- 0.006689293614962827)
Precision: f0.982110184008608 (+/- 0.0079894420717991)
Recall: f0.9545354974212689 (+/- 0.012840426979372226)
No confidence threshold - 666 classified
[[527  22]
 [ 16 101]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.97      0.96      0.86      0.97      0.91      0.84       549
          0       0.82      0.86      0.96      0.84      0.91      0.82       117

avg / total       0.94      0.94      0.88      0.94      0.91      0.83       666


Confidence threshold > 0.6 - 638 classified
[[517  16]
 [ 11  94]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.97      0.90      0.97      0.93      0.87       533
          0       0.85      0.90      0.97      0.87      0.93      0.86       105

avg / total       0.96      0.96      0.91      0.96      0.93      0.87       638


Confidence threshold > 0.7 - 596 classified
[[500   5]
 [  7  84]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.99      0.92      0.99      0.96      0.92       505
          0       0.94      0.92      0.99      0.93      0.96      0.91        91

avg / total       0.98      0.98      0.93      0.98      0.96      0.92       596


Confidence threshold > 0.8 - 542 classified
[[473   5]
 [  3  61]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.99      0.95      0.99      0.97      0.95       478
          0       0.92      0.95      0.99      0.94      0.97      0.94        64

avg / total       0.99      0.99      0.96      0.99      0.97      0.95       542


Confidence threshold > 0.9 - 477 classified
[[431   2]
 [  1  43]]
                   pre       rec       spe        f1       geo       iba       sup

          1       1.00      1.00      0.98      1.00      0.99      0.97       433
          0       0.96      0.98      1.00      0.97      0.99      0.97        44

avg / total       0.99      0.99      0.98      0.99      0.99      0.97       477
```

**Comment by Aftaab99** [comment]
**Tracking model**
i) Undersampling
```
Cross Validation scores:
Accuracy: f0.9324201167060628 (+/- 0.005342898438596523)
Precision: f0.34958073296676934 (+/- 0.01891190750406595)
Recall: f0.9398368247765838 (+/- 0.023110969568626436)
No confidence threshold - 4948 classified
[[ 173   11]
 [ 302 4462]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.36      0.94      0.94      0.53      0.94      0.88       184
          0       1.00      0.94      0.94      0.97      0.94      0.88      4764

avg / total       0.97      0.94      0.94      0.95      0.94      0.88      4948


Confidence threshold > 0.6 - 4733 classified
[[ 166    7]
 [ 214 4346]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.44      0.96      0.95      0.60      0.96      0.92       173
          0       1.00      0.95      0.96      0.98      0.96      0.91      4560

avg / total       0.98      0.95      0.96      0.96      0.96      0.91      4733


Confidence threshold > 0.7 - 4500 classified
[[ 160    4]
 [ 151 4185]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.51      0.98      0.97      0.67      0.97      0.94       164
          0       1.00      0.97      0.98      0.98      0.97      0.94      4336

avg / total       0.98      0.97      0.98      0.97      0.97      0.94      4500


Confidence threshold > 0.8 - 4200 classified
[[ 155    0]
 [ 100 3945]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.61      1.00      0.98      0.76      0.99      0.98       155
          0       1.00      0.98      1.00      0.99      0.99      0.97      4045

avg / total       0.99      0.98      1.00      0.98      0.99      0.97      4200


Confidence threshold > 0.9 - 3602 classified
[[ 126    0]
 [  56 3420]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.69      1.00      0.98      0.82      0.99      0.99       126
          0       1.00      0.98      1.00      0.99      0.99      0.98      3476

avg / total       0.99      0.98      1.00      0.99      0.99      0.98      3602
```
ii) InstanceHardnessThreshold
```
Cross Validation scores:
Accuracy: f0.938019443798481 (+/- 0.008766847024990242)
Precision: f0.5924586177003984 (+/- 0.03605995355971241)
Recall: f0.9849668946054487 (+/- 0.012570713035758664)
No confidence threshold - 4948 classified
[[ 171   13]
 [ 386 4378]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.31      0.93      0.92      0.46      0.92      0.85       184
          0       1.00      0.92      0.93      0.96      0.92      0.85      4764

avg / total       0.97      0.92      0.93      0.94      0.92      0.85      4948


Confidence threshold > 0.6 - 4805 classified
[[ 168   12]
 [ 324 4301]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.34      0.93      0.93      0.50      0.93      0.87       180
          0       1.00      0.93      0.93      0.96      0.93      0.87      4625

avg / total       0.97      0.93      0.93      0.95      0.93      0.87      4805


Confidence threshold > 0.7 - 4670 classified
[[ 163    9]
 [ 274 4224]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.37      0.95      0.94      0.54      0.94      0.89       172
          0       1.00      0.94      0.95      0.97      0.94      0.89      4498

avg / total       0.97      0.94      0.95      0.95      0.94      0.89      4670


Confidence threshold > 0.8 - 4518 classified
[[ 156    7]
 [ 225 4130]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.41      0.96      0.95      0.57      0.95      0.91       163
          0       1.00      0.95      0.96      0.97      0.95      0.91      4355

avg / total       0.98      0.95      0.96      0.96      0.95      0.91      4518


Confidence threshold > 0.9 - 4225 classified
[[ 145    5]
 [ 147 3928]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.50      0.97      0.96      0.66      0.97      0.93       150
          0       1.00      0.96      0.97      0.98      0.97      0.93      4075

avg / total       0.98      0.96      0.97      0.97      0.97      0.93      4225
```

File: bugbug/model.py [hunk 2]
```diff
@@ -70,26 +69,25 @@ def bugs():
...
         # Split dataset in training and test.
         X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
-
-        if self.undersampling_enabled:
-            # Under-sample the majority classes, as the datasets are imbalanced.
-            X_train, y_train = RandomUnderSampler(random_state=0).fit_sample(X_train, y_train)
-
-        print(f'X_train: {X_train.shape}, y_train: {y_train.shape}')
-        print(f'X_test: {X_test.shape}, y_test: {y_test.shape}')
+        pipeline = make_pipeline(self.sampler, self.clf)
```
Comment (marco-c) [inline bugbug/model.py]: Why do we need to make a pipeline with imblearn?

File: bugbug/model.py [hunk 3]
```diff
@@ -70,26 +69,25 @@ def bugs():
...
+            scores = cross_validate(pipeline, X_train, y_train, scoring=scorings, cv=5)
+
             print('Cross Validation scores:')
             for scoring in scorings:
                 score = scores[f'test_{scoring}']
                 print(f'{scoring.capitalize()}: f{score.mean()} (+/- {score.std() * 2})')
 
-        # Evaluate results on the test set.
+        # Training on the resampled dataset
+        X_train, y_train = self.sampler.fit_resample(X_train, y_train)
```
Comment (marco-c) [inline bugbug/model.py]: Is there a difference between `fit_sample` and `fit_resample`?

File: bugbug/model.py [hunk 4]
```diff
@@ -70,26 +69,25 @@ def bugs():
...
-            scores = cross_validate(self.clf, X_train, y_train, scoring=scorings, cv=5)
+            scores = cross_validate(pipeline, X_train, y_train, scoring=scorings, cv=5)
+
             print('Cross Validation scores:')
             for scoring in scorings:
                 score = scores[f'test_{scoring}']
                 print(f'{scoring.capitalize()}: f{score.mean()} (+/- {score.std() * 2})')
 
-        # Evaluate results on the test set.
+        # Training on the resampled dataset
```
Comment (marco-c) [inline bugbug/model.py]: Nit: point at the end of the sentence

File: bugbug/models/bug.py [hunk 5]
```diff
@@ -50,12 +51,14 @@ def __init__(self, lemmatization=False):
 
                 ('title', self.text_vectorizer(), 'title'),
 
-                ('first_comment', self.text_vectorizer(), 'first_comment'),
+                ('first_comment', self.text_vectorizer(stop_words='english', min_df=0.001), 'first_comment'),
```
Comment (marco-c) [inline bugbug/models/bug.py]: Why did you change these?

File: bugbug/models/tracking.py [hunk 6]
```diff
@@ -58,12 +59,14 @@ def __init__(self, lemmatization=False):
             ('union', ColumnTransformer([
                 ('data', DictVectorizer(), 'data'),
 
-                ('title', self.text_vectorizer(), 'title'),
+                ('title', self.text_vectorizer(stop_words='english', min_df=0.01), 'title'),
+
+                ('comments', self.text_vectorizer(stop_words='english', min_df=0.01), 'comments'),
```
Comment (marco-c) [inline bugbug/models/tracking.py]: Did you change these because of Out Of Memory issues? If so, `min_df=0.01` seems a little too small.
Please remove `stop_words`

**Comment by marco-c** [comment]
We should support models without a sampler too. Maybe define a no-op sampler by default in model.py if possible. Otherwise, make it None by default and don't apply it in `train` in case it is None.

File: bugbug/model.py (hunk 3 above)
Comment (Aftaab99) [reply bugbug/model.py]: It seems `fit_sample` is an alias for `fit_resample` for backward compatibility. It was called 'fit_sample` in earlier versions.

File: bugbug/model.py (hunk 2 above)
Comment (Aftaab99) [reply bugbug/model.py]: Making a pipeline makes sure oversampling occurs only on the training data and not on the validation data. Without a pipeline we'll need to oversample before passing the data to cross validate, which will lead to lot of duplicate/similar values ending up in the validation set, thus not giving a correct idea of how well the model is doing.
[https://beckernick.github.io/oversampling-modeling/](url) 
[https://stackoverflow.com/questions/50245684/using-smote-with-gridsearchcv-in-scikit-learn](url).

File: bugbug/models/tracking.py (hunk 6 above)
Comment (Aftaab99) [reply bugbug/models/tracking.py]: Yes, I used 0.01 because of memory issues. I tried with 0.05 and 0.001 both of which gave me the same errors. I've set it to 0.001 as you say its too small, but I can't test the model as it would require more than the 8GB of RAM I'm currently having.

File: bugbug/model.py [hunk 7]
```diff
@@ -70,26 +70,29 @@ def bugs():
...
-        if self.undersampling_enabled:
-            # Under-sample the majority classes, as the datasets are imbalanced.
-            X_train, y_train = RandomUnderSampler(random_state=0).fit_sample(X_train, y_train)
-
-        print(f'X_train: {X_train.shape}, y_train: {y_train.shape}')
-        print(f'X_test: {X_test.shape}, y_test: {y_test.shape}')
+        if self.sampler is not None:
+            pipeline = make_pipeline(self.sampler, self.clf)
+        else:
+            pipeline = make_pipeline(self.clf)
```
Comment (marco-c) [inline bugbug/model.py]: Could this simply be `pipeline = make_pipeline(self.clf)`?

Comment (marco-c) [reply bugbug/model.py]: Sorry, `pipeline = self.clf`

File: bugbug/models/bug.py [hunk 8]
```diff
@@ -50,12 +51,14 @@ def __init__(self, lemmatization=False):
 
                 ('title', self.text_vectorizer(), 'title'),
 
-                ('first_comment', self.text_vectorizer(), 'first_comment'),
+                ('first_comment', self.text_vectorizer(min_df=0.001), 'first_comment'),
+
+                ('comments', self.text_vectorizer(min_df=0.001), 'comments'),
```
Comment (marco-c) [inline bugbug/models/bug.py]: Let's remove the `min_df` changes (both here and in tracking.py), I'll set them myself by checking what we can use on my machine and on the AWS instance we are using.

**Comment by marco-c** [comment]
Looking good, just a couple of changes needed!

File: bugbug/models/bug.py [hunk 9]
```diff
@@ -53,9 +54,11 @@ def __init__(self, lemmatization=False):
                 ('first_comment', self.text_vectorizer(), 'first_comment'),
 
                 ('comments', self.text_vectorizer(), 'comments'),
+
```
Comment (marco-c) [inline bugbug/models/bug.py]: Remove newline here

File: bugbug/models/tracking.py [hunk 10]
```diff
@@ -61,9 +62,11 @@ def __init__(self, lemmatization=False):
                 ('title', self.text_vectorizer(), 'title'),
 
                 ('comments', self.text_vectorizer(), 'comments'),
+
```
Comment (marco-c) [inline bugbug/models/tracking.py]: And here

**Comment by marco-c** [comment]
One last thing, we should keep using RandomUnderSampler for the other models which were using it. undersampling_enabled used to be True by default, but now we have the opposite behavior (sampler being None by default).
So, for models which didn't have undersampling_enabled set to False and which are not now using oversampling, we should set sampler to RandomUnderSampler.

File: bugbug/model.py [hunk 11]
```diff
@@ -24,8 +25,8 @@ def __init__(self, lemmatization=False):
         else:
             self.text_vectorizer = TfidfVectorizer
 
-        self.undersampling_enabled = True
         self.cross_validation_enabled = True
+        self.sampler = RandomUnderSampler(random_state=0) if self.undersampling_enabled else self.sampler
```
Comment (marco-c) [inline bugbug/model.py]: No, make this None by default, and simply define it in each model as needed.
Remove the self.undersampling_enabled attribute.

File: bugbug/models/bug.py [hunk 12]
```diff
@@ -16,6 +17,8 @@
 
 class BugModel(Model):
     def __init__(self, lemmatization=False):
+        self.undersampling_enabled = False
```
Comment (marco-c) [inline bugbug/models/bug.py]: Remvoe the attribute

File: bugbug/models/component.py [hunk 13]
```diff
@@ -17,11 +17,9 @@
 
 class ComponentModel(Model):
     def __init__(self, lemmatization=False):
+        self.undersampling_enabled = True
```
Comment (marco-c) [inline bugbug/models/component.py]: Remove the attribute

File: bugbug/models/component.py [hunk 14]
```diff
@@ -17,11 +17,9 @@
...
         Model.__init__(self, lemmatization)
 
-        self.undersampling_enabled = False
-        self.cross_validation_enabled = False
```
Comment (marco-c) [inline bugbug/models/component.py]: Keep the cross_validation_enabled here

File: bugbug/models/devdocneeded.py [hunk 15]
```diff
@@ -15,6 +15,7 @@
 
 class DevDocNeededModel(Model):
     def __init__(self, lemmatization=False):
+        self.undersampling_enabled = True
```
Comment (marco-c) [inline bugbug/models/devdocneeded.py]: Remove this, and add sampler = RandomUnderSampler (after the Model.__init__ call)

Comment (marco-c) [reply bugbug/models/devdocneeded.py]: Same for qaneeded and uplift

File: bugbug/model.py [hunk 16]
```diff
@@ -70,26 +70,28 @@ def bugs():
...
+            scores = cross_validate(pipeline, X_train, y_train, scoring=scorings, cv=5)
+
             print('Cross Validation scores:')
             for scoring in scorings:
                 score = scores[f'test_{scoring}']
                 print(f'{scoring.capitalize()}: f{score.mean()} (+/- {score.std() * 2})')
 
-        # Evaluate results on the test set.
+        # Training on the resampled dataset if sampler is provided.
+        X_train, y_train = self.sampler.fit_resample(X_train, y_train)
```
Comment (marco-c) [inline bugbug/model.py]: Why did you modify here? It was good before :smile: 
You can't `fit_resample` for models for which the sampler is none

Comment (Aftaab99) [reply bugbug/model.py]: Oops! I'll fix that. But what about the Travis build failing due to out of memory errors? Should I add the the `min_df ` argument for the bug model back?

Comment (marco-c) [reply bugbug/model.py]: Don't worry, I'll take care of that myself ;)
